### DISCUSSION TITLE
How to extend Rule.NodeListeners when writing eslint plugin?

### DISCUSSION THREAD

**Discussion by Emptymu** [eslint/eslint#14785] [Q&A]
I'm writing an eslint plugin for a TypeScript + React project.
Take this `no-deprecated-component.ts` rule for an example:
<img width="598" alt="Screen Shot 2021-07-10 at 7 05 01 AM" src="https://user-images.githubusercontent.com/25358596/125143425-3dc53d80-e14d-11eb-840f-d2d52ed0df86.png">
`Rule.NodeListener` interface doesn't have node types for React. e.g. `JSXIdentifier`

What's the suggested way to get typing support for nodes like `JSXIdentifier`? Or it is not recommended to write rules in Typescript at all

**Comment by nzakas** [comment]
We don’t natively support TypeScript, so we don’t have a clean answer for you

**Reply by Emptymu** [reply]
Tried using `declare module` to merge declarations, but it didn't work. Because types like `type Node = Identifier | Expression | ...` can't be overridden. 

It would be nice if we can use generics to specify the AST node types and node listeners to work with. e.g. typescript-eslint/plugins's AST node types
